packages/grafana-ui/src/components/Logs/LogDetailsRow.tsx

```diff
@@ -0,0 +1,134 @@
+import React, { PureComponent } from 'react';
+import { LogRowModel, LogsParser, LogLabelStatsModel, calculateFieldStats } from '@grafana/data';
+
+import { Themeable } from '../../types/theme';
+import { withTheme } from '../../themes/index';
+import { getLogRowStyles } from './getLogRowStyles';
+
+//Components
+import { LogLabelStats } from './LogLabelStats';
+
+interface Props extends Themeable {
+  parsedValue: string;
+  parsedKey: string;
+  field?: string;
+  row?: LogRowModel;
+  canShowMetrics?: boolean;
+  canFilter?: boolean;
+  canFilterOut?: boolean;
+  parser?: LogsParser;
+  onClickFilterLabel?: (key: string, value: string) => void;
+  onClickFilterOutLabel?: (key: string, value: string) => void;
+  createStatsForLogs?: (field: string) => void;
+  getRows?: () => LogRowModel[];
+}
+
+interface State {
+  showFieldsStats: boolean;
+  fieldCount: number;
+  fieldLabel: string | null;
+  fieldStats: LogLabelStatsModel[] | null;
+  fieldValue: string | null;
+}
+
+class UnThemedLogDetailsRow extends PureComponent<Props, State> {
+  state: State = {
+    showFieldsStats: false,
+    fieldCount: 0,
+    fieldLabel: null,
+    fieldStats: null,
+    fieldValue: null,
+  };
+
+  filterLabel = () => {
+    const { onClickFilterLabel, parsedKey, parsedValue } = this.props;
+    if (onClickFilterLabel) {
+      onClickFilterLabel(parsedKey, parsedValue);
+    }
+  };
+
+  filterOutLabel = () => {
+    const { onClickFilterOutLabel, parsedKey, parsedValue } = this.props;
+    if (onClickFilterOutLabel) {
+      onClickFilterOutLabel(parsedKey, parsedValue);
+    }
+  };
+
+  showStats = () => {
+    if (!this.state.showFieldsStats) {
+      this.createStatsForLogs();
+    }
+    this.toggleFieldsStats();
+  };
+
+  toggleFieldsStats() {
+    this.setState(state => {
+      return {
+        showFieldsStats: !state.showFieldsStats,
+      };
+    });
+  }
+
+  createStatsForLogs() {
+    const { getRows, parser, field } = this.props;
+    if (field && getRows) {
+      const allRows = getRows();
+
+      // Build value-agnostic row matcher based on the field label
+      const fieldLabel = parser!.getLabelFromField(field);
```

Comment:
is parser optional? All the these ! operators make me feel that something is not correctly typed or checked.

Reply:
See the comment above. I will refactor LogDetailsRow and split it into 2 components that have clear props requirements and clear responsibilities. In that case, parser will be required and not optional. I will also look more closely on ! operators and see what's going on. I will keep you updated!

Reply:
Parsers are found dynamically per log row. If it returned one, then parsed fields are shown, which will have a button to calc the stats for the parsed field. If no parser is found, there are no parsed fields and this function cannot be triggered. Bit unclear how to model this in TS.